### PR TITLE
framework/ble_manager: implement ble client context

### PR DESCRIPTION
**Pull request by an4967**
- implement ble client context for multi connection

### CODE REVIEW

File: apps/examples/ble_rmc/ble_rmc_main.c [hunk 1]
```diff
@@ -67,27 +68,40 @@ static void ble_device_scanned_cb(ble_scanned_device *scanned_device)
...
+	RMC_LOG(RMC_TAG, "Conn MAC : %02x:%02x:%02x:%02x:%02x:%02x\n", 
+		dev->conn_info.addr.mac[0],
+		dev->conn_info.addr.mac[1],
+		dev->conn_info.addr.mac[2],
+		dev->conn_info.addr.mac[3],
+		dev->conn_info.addr.mac[4],
+		dev->conn_info.addr.mac[5]
+	);
+
+	ret = sem_post(&g_conn_sem);
```
Comment (jeongchanKim) [inline apps/examples/ble_rmc/ble_rmc_main.c:97]: The return of `sem_post` is not used anywhere.
So I think it is not needed or it is better to check the return.

File: framework/include/ble_manager/ble_client.h [hunk 2]
```diff
@@ -242,16 +241,17 @@ ble_result_e ble_client_operation_enable_notification(ble_client_operation_handl
  *   I/O operation with Server(Peripheral).
  * 
  * Input Parameters:
- *   handle  - This includes the handle ID and the attribute ID.
- *   data    - I/O buffer which trasfers data.
- *             (On read, this value should be allocated and set the length as the size of buffer.)
+ *   ctx          - The context of client.
+ *   attr_handle  - attribute handle.
+ *   data         - I/O buffer which trasfers data.
```
Comment (jeongchanKim) [inline framework/include/ble_manager/ble_client.h:246]: Please change `trasfers` to `transfers`.

File: apps/examples/ble_rmc/ble_rmc_main.c (hunk 1 above)
Comment (an4967) [reply apps/examples/ble_rmc/ble_rmc_main.c]: I fixed it

File: framework/include/ble_manager/ble_client.h (hunk 2 above)
Comment (an4967) [reply framework/include/ble_manager/ble_client.h]: I fixed it

File: apps/examples/ble_rmc/ble_rmc_main.c [hunk 3]
```diff
@@ -67,27 +68,41 @@ static void ble_device_scanned_cb(ble_scanned_device *scanned_device)
...
-	RMC_LOG(RMC_TAG, "'%s' is called\n", __FUNCTION__);
+	RMC_LOG(RMC_TAG, "'%s' is called[%p]\n", __FUNCTION__, ctx);
+	(void)sem_post(&g_conn_sem);
 	return;
 }
 
-static void ble_device_connected_cb(ble_device_connected *dev)
+static void ble_device_connected_cb(ble_client_ctx *ctx, ble_device_connected *dev)
 {
 	int ret;
```
Comment (jeongchanKim) [inline apps/examples/ble_rmc/ble_rmc_main.c:80]: Please remove unnecessary local variable :)

File: apps/examples/ble_rmc/ble_rmc_main.c [hunk 4]
```diff
@@ -339,29 +365,43 @@ int ble_rmc_main(int argc, char *argv[])
...
+			}
 
-		sleep(2);
+			sleep(2);
 
-		if (g_scan_done != 1) {
-			RMC_LOG(RMC_TAG, "scan fail to get adv packet\n", ret);
-			return 0;
+			if (g_scan_done != 1) {
+				RMC_LOG(RMC_TAG, "scan fail to get adv packet\n", ret);
```
Comment (jeongchanKim) [inline apps/examples/ble_rmc/ble_rmc_main.c:399]: `ret` is not used anywhere. Is it needed?

Comment (an4967) [reply apps/examples/ble_rmc/ble_rmc_main.c]: I fixed it

File: apps/examples/ble_rmc/ble_rmc_main.c (hunk 3 above)
Comment (an4967) [reply apps/examples/ble_rmc/ble_rmc_main.c]: I fixed it

**Review by jeongchanKim — APPROVED** [review]
LGTM

File: apps/examples/ble_rmc/ble_rmc_main.c [hunk 5]
```diff
@@ -101,28 +130,28 @@ static void ble_operation_notification_cb(ble_client_operation_handle *handle, b
 
 static void ble_server_connected_cb(ble_conn_handle con_handle, ble_server_connection_type_e conn_type, uint8_t mac[BLE_BD_ADDR_MAX_LEN])
 {
-	RMC_LOG(RMC_TAG, "'%s' is called\n", __FUNCTION__);
-	printf("conn : %d / conn_type : %d\n", con_handle, conn_type);
-	printf("conn mac : %02x:%02x:%02x:%02x:%02x:%02x\n", mac[0], mac[1], mac[2], mac[3], mac[4], mac[5]);
+	RMC_LOG(RMC_SERVER_TAG, "'%s' is called[%d]\n", __FUNCTION__);
```
Comment (jeongchanKim) [inline apps/examples/ble_rmc/ble_rmc_main.c:133]: You used `%s` and `%d`, but passed only `__FUNCTION__`.
In this case, trash value can be printed.
Please check the parameters.

Comment (an4967) [reply apps/examples/ble_rmc/ble_rmc_main.c]: You are right.
I fixed it

**Review by jeongchanKim — DISMISSED** [review]
LGTM

File: framework/include/ble_manager/ble_client.h [hunk 6]
```diff
@@ -144,78 +141,79 @@ ble_result_e ble_client_stop_scan(void);
...
+ * Name: ble_client_disconnect_all
  *
  * Description:
- *   Get the specific device info with conn_handle.
- *
- * Input Parameters:
- *   conn_handle          - The handle ID of connected device.
- *   out_connected_device - This should have pre-allocated and cleared memory.
- *                          The BLE driver will write the device info.
+ *   Disconnect all of the connected devicse.
```
Comment (abhishek-samsung) [inline framework/include/ble_manager/ble_client.h:174]: Please change devicse --> devices.

Comment (an4967) [reply framework/include/ble_manager/ble_client.h]: Ok, I fixed it

File: apps/examples/ble_rmc/ble_rmc_main.c [hunk 7]
```diff
@@ -329,39 +363,53 @@ int ble_rmc_main(int argc, char *argv[])
...
+			ret = ble_client_start_scan(&filter, &scan_config);
 
-		if (ret != BLE_MANAGER_SUCCESS) {
-			RMC_LOG(RMC_TAG, "scan start fail[%d]\n", ret);
-			return 0;
-		}
+			if (ret != BLE_MANAGER_SUCCESS) {
+				RMC_LOG(RMC_CLIENT_TAG, "scan start fail[%d]\n", ret);
+				return 0;
+			}
```
Comment (dineshdas-samsung) [inline apps/examples/ble_rmc/ble_rmc_main.c:402]: Can we please remove the extra tabs ?

File: apps/examples/ble_rmc/ble_rmc_main.c [hunk 8]
```diff
@@ -150,28 +180,37 @@ static ble_server_gatt_t gatt_profile[] = {
...
  ****************************************************************************/
 int ble_rmc_main(int argc, char *argv[])
 {
-	RMC_LOG(RMC_TAG, "test!!\n");
+	RMC_LOG(RMC_TAG, "- BLE Remote Test -\n");
 
 	ble_result_e ret = BLE_MANAGER_FAIL;
-
-	ble_client_init_config client_config = {
+
```
Comment (dineshdas-samsung) [inline apps/examples/ble_rmc/ble_rmc_main.c:203]: please remove the extra tabs.

File: apps/examples/ble_rmc/ble_rmc_main.c (hunk 7 above)
Comment (an4967) [reply apps/examples/ble_rmc/ble_rmc_main.c:402]: These are not extra tabs.
Checking 'ret' value is needed in 'else' state.

File: apps/examples/ble_rmc/ble_rmc_main.c (hunk 8 above)
Comment (an4967) [reply apps/examples/ble_rmc/ble_rmc_main.c]: Ok, I removed them

**Review by pillip8282 — APPROVED** [review]
lgtm

**Review by pillip8282 — APPROVED** [review]
lgtm

**Pull request merged**